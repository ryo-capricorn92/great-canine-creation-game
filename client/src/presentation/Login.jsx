import React from 'react';
import styled from 'styled-components';

const Container = styled.div`
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

const Label = styled.label`
  font-size: 15px;
  font-weight: bold;
  margin-right: 5px;
`;

const LabelCell = styled.td`
  padding: 3px;
  text-align: right;
`;

const InputCell = styled.td`
  position: relative;
  padding: 3px;
`;

const SubmitCell = styled.td.attrs({
  colSpan: '2',
})`
  padding: 5px;
`;

const Error = styled.div`
  color: red;
  background-color: #e5cece;
  border-radius: 3px;
  padding: 5px;
  margin-top: 10px;
`;

class Login extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      message: '',
    };
  }

  render() {
    return (
      <Container>
        <form onSubmit={this.onSubmit}>
          <table>
            <tbody>
              <tr>
                <LabelCell>
                  <Label htmlFor="username">Username:</Label>
                </LabelCell>
                <InputCell>
                  <input name="username" required />
                </InputCell>
              </tr>
              <tr>
                <LabelCell>
                  <Label htmlFor="password">Password:</Label>
                </LabelCell>
                <InputCell>
                  <input name="password" type="password" required />
                </InputCell>
              </tr>
              <tr>
                <SubmitCell>
                  <input type="submit" value="Submit" />
                </SubmitCell>
              </tr>
            </tbody>
          </table>
        </form>
        {this.state.message ? (<Error>{this.state.message}</Error>) : undefined}
      </Container>
    );
  }
}

export default Login;
