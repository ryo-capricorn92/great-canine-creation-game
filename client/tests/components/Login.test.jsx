import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import fetch from 'jest-fetch-mock';

import Login from '../../src/presentation/Login';

global.fetch = fetch;

const submitForm = {
  username: {
    value: 'Ryo',
  },
  password: {
    value: 'testingPassword123',
  },
};

describe('Login', () => {
  it('should render Login without failing', () => {
    const wrapper = mount(<Login />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should "submit" a valid form properly', (done) => {
    const wrapper = mount(<Login />);
    const onSubmitSpy = jest.spyOn(wrapper.instance(), 'onSubmit');
    wrapper.update();

    fetch.mockRejectOnce(JSON.stringify({
      ok: true,
      json: () => {},
    }));

    wrapper.find('form').simulate('submit', { target: submitForm });
    setImmediate(() => {
      expect(onSubmitSpy).toHaveBeenCalled();
      onSubmitSpy.mockReset();
      onSubmitSpy.mockRestore();
      expect(toJson(wrapper)).toMatchSnapshot();
      done();
    });
  });

  it('should alert the user if the username or password is wrong', (done) => {
    const wrapper = mount(<Login />);

    fetch.mockResponseOnce(JSON.stringify({
      ok: true,
      statusText: 'failure',
      status: '401',
      json: () => {},
    }));

    wrapper.find('form').simulate('submit', { target: submitForm });
    setImmediate(() => {
      expect(toJson(wrapper)).toMatchSnapshot();
      done();
    });
  });
});
