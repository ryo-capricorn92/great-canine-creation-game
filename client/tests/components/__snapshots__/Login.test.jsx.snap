// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should "submit" a valid form properly 1`] = `
<Login>
  <styled.div>
    <div
      className="sc-bdVaJa iBeQiE"
    >
      <form
        onSubmit={[Function]}
      >
        <table>
          <tbody>
            <tr>
              <styled.td>
                <td
                  className="sc-htpNat gZOKTT"
                >
                  <styled.label
                    htmlFor="username"
                  >
                    <label
                      className="sc-bwzfXH fGBoem"
                      htmlFor="username"
                    >
                      Username:
                    </label>
                  </styled.label>
                </td>
              </styled.td>
              <styled.td>
                <td
                  className="sc-bxivhb iydZbq"
                >
                  <input
                    name="username"
                    required={true}
                  />
                </td>
              </styled.td>
            </tr>
            <tr>
              <styled.td>
                <td
                  className="sc-htpNat gZOKTT"
                >
                  <styled.label
                    htmlFor="password"
                  >
                    <label
                      className="sc-bwzfXH fGBoem"
                      htmlFor="password"
                    >
                      Password:
                    </label>
                  </styled.label>
                </td>
              </styled.td>
              <styled.td>
                <td
                  className="sc-bxivhb iydZbq"
                >
                  <input
                    name="password"
                    required={true}
                    type="password"
                  />
                </td>
              </styled.td>
            </tr>
            <tr>
              <styled.td>
                <td
                  className="sc-ifAKCX gYKwRG"
                  colSpan="2"
                >
                  <input
                    type="submit"
                    value="Submit"
                  />
                </td>
              </styled.td>
            </tr>
          </tbody>
        </table>
      </form>
    </div>
  </styled.div>
</Login>
`;

exports[`Login should alert the user if the username or password is wrong 1`] = `
<Login>
  <styled.div>
    <div
      className="sc-bdVaJa iBeQiE"
    >
      <form
        onSubmit={[Function]}
      >
        <table>
          <tbody>
            <tr>
              <styled.td>
                <td
                  className="sc-htpNat gZOKTT"
                >
                  <styled.label
                    htmlFor="username"
                  >
                    <label
                      className="sc-bwzfXH fGBoem"
                      htmlFor="username"
                    >
                      Username:
                    </label>
                  </styled.label>
                </td>
              </styled.td>
              <styled.td>
                <td
                  className="sc-bxivhb iydZbq"
                >
                  <input
                    name="username"
                    required={true}
                  />
                </td>
              </styled.td>
            </tr>
            <tr>
              <styled.td>
                <td
                  className="sc-htpNat gZOKTT"
                >
                  <styled.label
                    htmlFor="password"
                  >
                    <label
                      className="sc-bwzfXH fGBoem"
                      htmlFor="password"
                    >
                      Password:
                    </label>
                  </styled.label>
                </td>
              </styled.td>
              <styled.td>
                <td
                  className="sc-bxivhb iydZbq"
                >
                  <input
                    name="password"
                    required={true}
                    type="password"
                  />
                </td>
              </styled.td>
            </tr>
            <tr>
              <styled.td>
                <td
                  className="sc-ifAKCX gYKwRG"
                  colSpan="2"
                >
                  <input
                    type="submit"
                    value="Submit"
                  />
                </td>
              </styled.td>
            </tr>
          </tbody>
        </table>
      </form>
    </div>
  </styled.div>
</Login>
`;

exports[`Login should render Login without failing 1`] = `
<Login>
  <styled.div>
    <div
      className="sc-bdVaJa iBeQiE"
    >
      <form
        onSubmit={[Function]}
      >
        <table>
          <tbody>
            <tr>
              <styled.td>
                <td
                  className="sc-htpNat gZOKTT"
                >
                  <styled.label
                    htmlFor="username"
                  >
                    <label
                      className="sc-bwzfXH fGBoem"
                      htmlFor="username"
                    >
                      Username:
                    </label>
                  </styled.label>
                </td>
              </styled.td>
              <styled.td>
                <td
                  className="sc-bxivhb iydZbq"
                >
                  <input
                    name="username"
                    required={true}
                  />
                </td>
              </styled.td>
            </tr>
            <tr>
              <styled.td>
                <td
                  className="sc-htpNat gZOKTT"
                >
                  <styled.label
                    htmlFor="password"
                  >
                    <label
                      className="sc-bwzfXH fGBoem"
                      htmlFor="password"
                    >
                      Password:
                    </label>
                  </styled.label>
                </td>
              </styled.td>
              <styled.td>
                <td
                  className="sc-bxivhb iydZbq"
                >
                  <input
                    name="password"
                    required={true}
                    type="password"
                  />
                </td>
              </styled.td>
            </tr>
            <tr>
              <styled.td>
                <td
                  className="sc-ifAKCX gYKwRG"
                  colSpan="2"
                >
                  <input
                    type="submit"
                    value="Submit"
                  />
                </td>
              </styled.td>
            </tr>
          </tbody>
        </table>
      </form>
    </div>
  </styled.div>
</Login>
`;
